jormungandr/src/blockchain/bootstrap.rs

```diff
@@ -0,0 +1,135 @@
+use super::tip::TipUpdater;
+use crate::blockcfg::{Block, HeaderHash};
+use crate::blockchain::{
+    chain::{CheckHeaderProof, StreamInfo, StreamReporter},
+    Blockchain, Ref, Tip,
+};
+use crate::metrics::Metrics;
+use chain_core::property::Deserialize;
+use chain_network::data as net_data;
+use chain_network::error::Error as NetworkError;
+use futures::prelude::*;
+use tokio_util::sync::CancellationToken;
+
+use std::sync::Arc;
+
+#[derive(Debug, thiserror::Error)]
+pub enum Error {
+    #[error(transparent)]
+    BlockchainError(#[from] super::Error),
+    #[error("received block {0} is not connected to the block chain")]
+    BlockMissingParent(HeaderHash),
+    #[error("bootstrap pull stream failed")]
+    PullStreamFailed(#[source] NetworkError),
+    #[error("failures while deserializing block from stream")]
+    BlockDeserialize(#[from] std::io::Error),
+    #[error("the bootstrap process was interrupted")]
+    Interrupted,
+}
+
+pub async fn bootstrap_from_stream<S>(
+    blockchain: Blockchain,
+    branch: Tip,
+    stream: S,
+    cancellation_token: CancellationToken,
+) -> Result<Option<Arc<Ref>>, Error>
+where
+    S: Stream<Item = Result<net_data::Block, NetworkError>> + Unpin,
+{
+    let block0 = *blockchain.block0();
+    let mut tip_updater = TipUpdater::new(
+        branch,
+        blockchain.clone(),
+        None,
+        None,
+        Metrics::builder().build(),
+    );
+
+    let mut bootstrap_info = StreamReporter::new(report);
+    let mut maybe_parent_tip = None;
+
+    // This stream will either end when the block stream is exhausted or when
+    // the cancellation signal arrives. Building such stream allows us to
+    // correctly write all blocks and update the block tip upon the arrival of
+    // the cancellation signal.
+    let cancel = cancellation_token.cancelled().boxed();
+    let mut stream = stream
+        .map_err(Error::PullStreamFailed)
+        .map(|maybe_block| maybe_block.and_then(|b| Ok(Block::deserialize(b.as_bytes())?)))
```

Comment:
```suggestion
        .map(|maybe_block| maybe_block.and_then(|b| Block::deserialize(b.as_bytes()).map_err(Error::BlockDeserialize)))
```

Reply:
I think I'm missing something :sweat_smile:, what is the advantage?

Reply:
Just looks better/more conventional IMO.